tests/System.Slices.Tests/MemoryTests.cs

```diff
@@ -108,41 +110,111 @@ public void RacyAccess()
             }
         }
 
-        [Fact]
-        public unsafe void ReferenceCounting()
+
+        #region helpers
+        private static object[] MakeArray(params object[] args)
         {
-            var owned = new CustomMemory();
-            var memory = owned.Memory;
-            Assert.Equal(0, owned.ReferenceCountChangeCount);
-            Assert.Equal(0, owned.ReferenceCount);
-            using (memory.Reserve()) {
-                Assert.Equal(1, owned.ReferenceCountChangeCount);
-                Assert.Equal(1, owned.ReferenceCount);
+            return args;
+        }
+
+        /// Generate the ith permutation of the argss array of arrays.
+        private static object[] MakeIthPermutation(int i, object[][] argss) {
+            var res = new object[argss.Length];
+            for (int j = 0; j < argss.Length; j++) {
+                var arg = argss[j];
+                res[j] = arg[i % arg.Length];
+                i /= arg.Length;
             }
-            Assert.Equal(2, owned.ReferenceCountChangeCount);
-            Assert.Equal(0, owned.ReferenceCount);
+            return res;
+        }
+        private static IEnumerable<object[]> MakePermutations(params object[][] argss)
+        {
+            if (argss.Length != 0) {
+                var size = 1;
+                // Calculate number of permutations.
+                foreach (var args in argss) {
+                    size *= args.Length;
+                }
+                object[][] ress = new object[size][];
+                for (int i = 0; i < size; i++) {
+                    ress[i] = MakeIthPermutation(i, argss);
+                }
+                return ress;
+            }
+            return null;
+        }
+        #endregion
+
+        public static IEnumerable<object[]> ReservationPerformanceData =>
+            MakePermutations(MakeArray(1,4,10),
+                             MakeArray(100,1000),
+                             MakeArray(1,2,4,8,12),
+                             MakeArray(Mode.None, Mode.RefCount, Mode.LocalRefCount));
+
+
+        [MemberData(nameof(ReservationPerformanceData))]
+        [Benchmark]
```

Comment:
Re your question about the project for the tests:

I think it's ok for these tests to be here or in the Benchmarks project. I would not create a separate test project for these.